File: Programmers/Level 1/키패드 누르기/Solution.py
def solution(numbers, hand):
    answer = ''    
    left = '*'
    right = '#'
    
    for n in numbers:
        if n in [1, 4, 7]:
            answer += 'L'
            left = n
        elif n in [3, 6, 9]:
            answer += 'R'
            right = n
        else:
            d_left = get_distance(left, n)
            d_right = get_distance(right, n)

            if d_left == d_right:
                if hand == "right":
                    answer += 'R'
                    right = n
                else:
                    answer += 'L'
                    left = n
            if d_left > d_right:
                answer += 'R'
                right = n
            if d_left < d_right:
                answer += 'L'
                left = n
    
    return answer

def get_distance(cur, target):
    keypads = {1: [0, 0], 2: [0, 1], 3: [0, 2],
              4: [1, 0], 5: [1, 1], 6: [1, 2],
              7: [2, 0], 8: [2, 1], 9: [2, 2],
              '*': [3, 0], 0: [3, 1], '#': [3, 2]}
    
    cur_pos = keypads[cur]
    target_pos = keypads[target]
    
    return abs(cur_pos[0]-target_pos[0]) + abs(cur_pos[1]-target_pos[1])

File: Programmers/Level 1/키패드 누르기/Solution.cpp
#include <string>
#include <vector>

using namespace std;

int distance(int cur, int target){
    int d = abs(target-cur);
    
    if(d==1||d==3)
        d = 1;
    else if(d==2||d==4||d==6)
        d = 2;
    else if(d==5||d==7||d==9)
        d = 3;
    else if(d==8||d==10)
        d = 4;
    
    return d;
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    int right = 10, left = 12;
    
    for(int i=0;i<numbers.size();i++){
        if(numbers[i]==0)
            numbers[i]=11;
        
        if(numbers[i]==1||numbers[i]==4||numbers[i]==7){
            left = numbers[i];
            answer += 'L';
        }
        else if(numbers[i]==3||numbers[i]==6||numbers[i]==9){
            right = numbers[i];
            answer += 'R';
        }
        else{
            int rightd = distance(right,numbers[i]);
            int leftd = distance(left,numbers[i]);
            
            if(rightd>leftd){
                left = numbers[i];
                answer += 'L';
            }
            else if(rightd<leftd){
                right = numbers[i];
                answer += 'R';
            }
            else{
                if(hand == "left"){
                    left = numbers[i];
                    answer += 'L';
                }
                else if(hand == "right"){
                    right = numbers[i];
                    answer += 'R';
                }
            }
        }
                
    }
    return answer;
}